### PR TITLE
Reducing test and validation time 

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

The tensornet backends' unit tests are currently quite slow. This affects developers running the full test suite, e.g., with `ctest`. 

**Note**: we expect tensornet backends don't perform well in these small circuits due to overheads associated with setting the tensor network, optimizing the contraction paths, etc. 

This PR identifies bottlenecks in our test suite to reduce the test time, especially for tensornet backends:

(1) A few VQE test cases use L-BFGS optimizer. There is a parameter to control the max number of line search trials, which we haven't exposed yet. Hence, expose this option and use it in our test to achieve faster convergence/shorter test time.

(2) The `gate_library_test` ran an exhaustive unitary verification of our kernel library. Since the state-vector backends should have covered this verification, reduce the test load on tensornet backends for this particular test (only check a few points).

(3) Removed an unnecessary run of non-gradient based optimization in the gradient tester. Rather, tighten the run and check conditions of the gradient tests (verifying that we can achieve good convergence with fewer iterations).

(4) Reduced the number of shots in tests that we're expecting simple distributions.

Resolved https://github.com/NVIDIA/cuda-quantum/issues/885

### CODE REVIEW

**Review by bettinaheim — APPROVED** [review]
lgtm

**Review by bmhowe23 — APPROVED** [review]
👍

File: docs/sphinx/examples/cpp/basics/cuquantum_backends.cpp [hunk 1]
```diff
@@ -28,7 +28,7 @@ struct ghz {
 };
 
 int main() {
-  auto counts = cudaq::sample(ghz{}, 28);
+  auto counts = cudaq::sample(/*shots*/ 100, ghz{}, 28);
```
Comment (schweitzpgi) [inline docs/sphinx/examples/cpp/basics/cuquantum_backends.cpp:31]:
```suggestion
  auto counts = cudaq::sample(/*shots=*/ 100, ghz{}, 28);
```
nit: silly LLVM style thing.

**Review by schweitzpgi — APPROVED** [review]
👍

**Pull request merged**